File: points.py
import numpy as np
from sklearn.cluster import KMeans


# class Point:
#     def __init__(self, value, weight=1):
#         self.value = value
#         self.dimension = len(value)
#         self.weight = weight


class Points:
    def __init__(self, size, dimension, seed=16):
        self.size = size
        self.dimension = dimension
        self.values = np.zeros((size, dimension))
        self.weights = np.ones(size)  # initialize all points weight equal to 1
        self.seed = seed

    def __len__(self):
        return self.size

    def set_seed(self, new_seed):
        self.seed = new_seed

    def fill_points(self, values, weights=None):
        assert values.shape == self.values.shape, 'Please input values with the shape of {}'.format(self.values.shape)
        self.values = values
        if weights is not None:
            assert weights.shape == self.weights.shape, 'Please input weights with the shape of {}'.format(
                self.weights.shape)
            self.set_weights(weights)

    def add_points(self, values, weights=None):
        assert self.values.shape[1] == values.shape[
            1], 'Please add the points with same dimension {} as current points'.format(self.values.shape[1])

        self.values = np.concatenate([self.values, values], axis=0)
        if weights is not None:
            assert len(weights) == len(values), 'The new values and weights are not in the same length.'
            self.weights = np.concatenate([self.weights, weights], axis=0)
        else:
            self.weights = np.concatenate([self.weights, np.ones(len(values))], axis=0)

        self.size += len(values)

    def set_weights(self, weights):
        assert len(weights) == len(self.values), 'The new weights are not in the same length with the values.'
        self.weights = weights

    def get_values(self):
        return self.values

    def get_dimension(self):
        return self.dimension

    def get_weights(self):
        return self.weights

    def kmeans_clustering(self, k):
        kmeans = KMeans(n_clusters=k, random_state=self.seed).fit(self.values, sample_weight=self.weights)
        return kmeans.cluster_centers_, kmeans.inertia_

File: cuda/points.cpp
//
// Created by Xiaobo Wu on 2021/1/18.
//

#include "points.h"
using namespace std;

namespace coreset {

    Points::Points() {
        this->size = 0;
        this->dimension = 0;
        this->values = vector<vector<float> >();
        this->weights = vector<float>();
    }

    Points::Points(size_int size, unsigned int dimension) {
        this->size = size;
        this->dimension = dimension;
        this->values = vector<vector<float> >(size, vector<float>(dimension));
        this->weights = vector<float>(size);
    }

    Points::Points(size_int size, unsigned int dimension, vector<vector<float> > values) {
        this->size = size;
        this->dimension = dimension;
        if (this->size != values.size())
            throw "values size is not matched setting size!";
        this->values = values;
        this->weights = vector<float>(size);
    }
    
    Points::Points(size_int size, unsigned int dimension, vector<vector<float> > values, vector<float> weights) {
        this->size = size;
        this->dimension = dimension;
        if (this->size != values.size())
            throw "values size is not matched setting size!";
        this->values = values;
        if (this->size != weights.size())
            throw "Weights size is not matched setting size!";
        this->weights = weights;
    }
    
    /*
    Points::Points(size_int size, unsigned int dimension, float* values, float* weights) {
        this->size = size;
        this->dimension = dimension;
	this->values = vector<vector<float> >(size, vector<float>(dimension));
        this->weights = vector<float>(weights, weights + size);
	
	//transfor the data from arrays to vector form
	for(int i=0; i<size; i++){
	    size_int start_id = i * dimension;
	    for(int j=0; j<dimension; j++){
	        this.values[i][j] = values[start_id + j];
	    }
	}
        
    }*/


    unsigned long int Points::Size() {
        return this->size;
    }

    unsigned int Points::Dimension() {
        return this->dimension;
    }

    void Points::FillPoints(vector<vector<float> > values, vector<float> weights) {
        if (this->size != values.size() && this->dimension != values[0].size())
            throw ("Please input the values with the shape of %d,%d.", this->size, this->dimension);
        this->values = values;
        if (weights.empty()) {
            this->weights = vector<float>(this->size, 1.0);
        } else if (this->size != weights.size()) {
            throw ("Please input the weights with the length of %d.", this->size);
        }
        else{
            this->weights = weights;
        }
    }
    
    void Points::AddPoints(vector<vector<float> > values, vector<float> weights) {
	if (this->values.empty())  //Right now, it's empty in the points object
	    this->dimension = values[0].size();

        if (this->dimension != values[0].size())
            throw ("Please add the points with same dimension %d as current points", this->dimension);

        this->values.insert(this->values.end(), values.begin(), values.end());
        if (weights.empty()){
            this->weights.insert(this->weights.end(), values.size(), 1.0);
        }
        else if (values.size() != weights.size()){
            throw "The new values and weights are not in the same length.";
        }
        else{
            this->weights.insert(this->weights.end(), weights.begin(), weights.end());
        }

        this->size += values.size();
    }

    void Points::SetWeights(vector<float> weights) {
        if (weights.size() != this->size){
            throw "The new weights are not in the same length with the values.";
        }
        this->weights = weights;
    }

    vector<vector<float> > Points::GetValues() {
        return this->values;
    }

    vector<float> Points::GetWeights() {
        return this->weights;
    }

}









